src/models/state/wallet/wallet_state.rs

```diff
@@ -1693,37 +1694,49 @@ impl WalletState {
                 continue;
             }
 
-            // Don't use inputs that are already spent by txs in mempool.
-            let absolute_index_set =
-                membership_proof.compute_indices(Tip5::hash(&wallet_status_element.utxo));
-            if index_sets_of_inputs_in_mempool_txs.contains(&absolute_index_set) {
-                continue;
-            }
-
+            // Don't use inputs that we can't spend
             let Some(spending_key) = self.find_spending_key_for_utxo(&wallet_status_element.utxo)
             else {
-                let utxo = &wallet_status_element.utxo;
-                warn!("spending key not found for utxo: {utxo:?}");
+                warn!(
+                    "spending key not found for utxo: {:?}",
+                    wallet_status_element.utxo
+                );
                 continue;
             };
 
-            input_funds.push(UnlockedUtxo::unlock(
+            // Create the transaction input object
+            let unlocked_utxo = UnlockedUtxo::unlock(
                 wallet_status_element.utxo.clone(),
                 spending_key,
                 membership_proof.clone(),
-            ));
-            allocated_amount =
-                allocated_amount + wallet_status_element.utxo.get_native_currency_amount();
+            );
+
+            // Don't use inputs that are already spent by txs in mempool.
+            total_available_amount_ignoring_mempool =
+                total_available_amount_ignoring_mempool + utxo_amount;
+            let absolute_index_set =
+                membership_proof.compute_indices(Tip5::hash(&wallet_status_element.utxo));
+            if index_sets_of_inputs_in_mempool_txs.contains(&absolute_index_set) {
+                continue;
+            }
+
+            // Select the input
+            input_funds.push(unlocked_utxo);
+            allocated_amount = allocated_amount + utxo_amount;
         }
 
-        // Sanity check. Shouldn't be possible to hit because of above balance
-        // check that also takes mempool into account.
-        assert!(
-            allocated_amount >= total_spend,
-            "UTXO allocation failed. This should not be possible. Requested: {}, Available: {}",
-            total_spend,
-            confirmed_available_amount_without_mempool_spends,
-        );
+        // If there aren't enough funds, catch and report error gracefully
+        if allocated_amount < total_spend {
+            bail!(
```

Comment:
Maybe, but that's definitely personal preference and not an outright recommendation:

```rust
            bail!(
                "UTXO allocation failed.\n\
                Requested: {total_spend}\n\
                Available, ignoring mempool: {total_available_amount_ignoring_mempool}\n\
                Allocated: {allocated_amount}",
            )
```